### DISCUSSION TITLE
Issue with the "scope" in OpenModelica

### DISCUSSION THREAD

**Discussion by ejperezHI** [OpenModelica/OpenModelica#11622] [Q&A]
Hello,

What is need to be done with the OpenModelica “scope” in order to work properly with the models? I have come across with this issue in my other problem (#11531) and recently in some more. I have reached the conclusion that has something to be with the OpenModelica's "scope". How do I manage properly the OpenModelica “scope” in order to be capable of exporting OpenModelica models as FMUs etc?

Is it really necessary to put the OpenModelica ‘.mo’ model inside the directory ‘.../.openmodelica/libraries/’? Or is not necessary?

The error I get after this exact code:
```python
import pyfmi
import pythonfmu
from OMPython import OMCSessionZMQ
omc = OMCSessionZMQ()
print(omc.sendExpression("getVersion()"))
omc.sendExpression("loadModel(Modelica)")
omc.sendExpression("getErrorString()")
omc.sendExpression("loadFile(\"/home/ejperez/Escritorio/OTROS_PROYECTOS/probandoOMG/EjemploOM_copia.mo\")")
omc.sendExpression("getErrorString()")
omc.sendExpression("setCommandLineOptions(\"--fmuRuntimeDepends=none\")")
omc.sendExpression("getErrorString()")
omc.sendExpression("buildModelFMU(EjemploOM_copia)") #This is the only command that returns a 'False'. The other ones return perfectly a 'True'.
omc.sendExpression("getErrorString()")
```

Is this one:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/a668d8dc-a57d-4eed-8411-b87258a8b876)

Which refers to the ‘scope’ of the OpenModelica, like something is wrong trying to work with an OpenModelica model (EjemploOM_copia.mo) that is not in the Modelicapath.

Checking the modelicapath in the OpenModelica itself (in the OpenModelica Terminal):
`getModelicaPath()`
I get the following path:

> /home/ejperez/.openmodelica/libraries/

Is the scope okay? Do I have to change the scope? What can I do to solve my issue?

Thanks in advance,

**Comment by AnHeuermann** [comment]
> Is it really necessary to put the OpenModelica ‘.mo’ model inside the directory ‘.../.openmodelica/libraries/’? Or is not necessary?

No. If OpenModelica can't find the class identifier in its list of loaded classes (added by `loadFile` or `loadModel`) it will try to load the class from the default library location. That means you can simulate e.g. `Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum` without first loading the Modelica standard library.

As pointed out in https://github.com/OpenModelica/OpenModelica/discussions/11531 the `loadFile` function is failing for you. This can have different reasons, but most likely your model is wrong or your path is wrong.

> What can I do to solve my issue?

You can try to load the file with OMEdit and see if it is telling you what the issue is.

Please try if you can get your workflow to run with a model that is known to work, e.g. [BouncingBall.mo](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Examples/BouncingBall.mo).

**Reply by ejperezHI** [reply]
Thanks for your reply,

> As pointed out in https://github.com/OpenModelica/OpenModelica/discussions/11531 the loadFile function is failing for you.

No. In my example, `loadFile` works "perfectly fine". At least, it returns a "True". What does return a 'False' and what originates the error posted in the image above is the `buildModelFmu` command.



> As pointed out in https://github.com/OpenModelica/OpenModelica/discussions/11531 the loadFile function is failing for you. This can have different reasons, but most likely your model is wrong or your path is wrong.

The model was made by a colleague expert in OpenModelica and it works perfectly (and it exports perfectly to FMU) **according to the people** that downloaded it and worked with it in #11531. So I guess that the model itself doesn't have any errors. In the other hand, the path of the script and the path of the model is the same, they are in the same directory. Here comes then my doubt..is it the 'scope' okay? Do I have to do anything with it?



> Please try if you can get your workflow to run with a model that is known to work, e.g. [BouncingBall.mo](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Examples/BouncingBall.mo).

I just downloaed it and proved my code (posted above in my forst message) with it. Althogh the `buildModelFMU` command returns also a 'False', it does work (magically) and creates the .fmu model. However, it doesn't give me the 'scope' error posted above.
After trying this, it does increase very greatly my doubts. 

- Why does this (BouncingBall.mo) work if the buildModel command returns a 'False'?
- Why my 'EjemploOM_copia' also returns a 'False' but it doesn't work?
- Why after running the exact same code, my 'EjemploOM_copia' returns the 'scope' error posted above and after running the 'BouncingBall.mo' it doesn't?


Thanks in advance,
